src/librustc_trans/abi.rs

```diff
@@ -866,11 +866,19 @@ impl<'a, 'tcx> FnType<'tcx> {
 
         if abi == Abi::Rust || abi == Abi::RustCall ||
            abi == Abi::RustIntrinsic || abi == Abi::PlatformIntrinsic {
-            let fixup = |arg: &mut ArgType<'tcx>| {
+            let fixup = |arg: &mut ArgType<'tcx>, is_ret: bool| {
                 if arg.is_ignore() { return; }
 
                 match arg.layout.abi {
                     layout::Abi::Aggregate { .. } => {}
+                    layout::Abi::Vector { .. } if abi != Abi::PlatformIntrinsic => {
+                        if is_ret {
+                            arg.make_indirect();
+                        } else {
+                            arg.make_indirect_byval();
```

Comment:
Do we want this? If so we should also use it for other indirect arguments below. But I think that if you already have a memory location to pass in, `byval`will make an extra stack copy instead of reusing the original.

Reply:
Oh hm so this may or may not be correct, I sort of just assumed that it was naively!

I did find though that this is what clang does, it marks all SIMD arguments as `byval` along with a pointer indirection behind them.

Reply:
I'm not sure I understand, Clang often passes vectors as immediates? see https://godbolt.org/g/2cSZJD

Clang may be using byval in some circumstances where the ABI dictates it (I believe byval has some ABI implications that differ from just passing a pointer), but for the "Rust" ABI we have more leeway.

Reply:
`byval` doesn't change correctness (as long as you're consistent), it's just a (platform) ABI choice, and we just don't do it for the Rust ABI. AFAIK it means "we have a pointer at the IR level but the data behind the pointer is passed on the stack" vs the default "this is a pointer like any other passed in a register (or the stack if we ran out of registers - but note that it's still just the pointer itself passed on the stack, not what it points to)".

Reply:
@rkruppe oh interesting! I wonder if this is an integer/float difference? I've been using https://godbolt.org/g/t4qzjA as testing ground which shows using the byval attribute...

Maybe it's 128/256 ABI? Unsure.

Reply:
@eddyb ok cool, I'll switch to just `make_indirect` for now.